// @fhevm/hardhat-plugin/_cjs/internal/FhevmEnvironmentPaths.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FhevmEnvironmentPaths = void 0;
exports.resolveFromConsumer = resolveFromConsumer;
const mock_utils_1 = require("@fhevm/mock-utils");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const resolve = __importStar(require("resolve"));
const constants_1 = __importDefault(require("../constants"));
const error_1 = require("../error");
const path_1 = require("./utils/path");
class FhevmEnvironmentPaths {
    constructor(root) {
        Object.defineProperty(this, "_root", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this._root = root;
        this._checkMockUtilsVersions();
    }
    get root() {
        return this._root;
    }
    get nodeModules() {
        return path.join(this.root, "node_modules");
    }
    get cache() {
        return path.join(this.root, "fhevmTemp");
    }
    get cachePrecompiledFhevmCoreContractsAddressesJson() {
        return path.join(this.cache, "precompiled-fhevm-core-contracts-addresses.json");
    }
    get cacheFhevmSolidityConfig() {
        return path.join(this.cache, path.join(constants_1.default.FHEVM_SOLIDITY_PACKAGE.name, path.dirname(constants_1.default.FHEVM_SOLIDITY_PACKAGE.configFile)));
    }
    get relCacheFhevmSolidityConfig() {
        const abs = this.cacheFhevmSolidityConfig;
        return path.relative(this.root, abs);
    }
    get relCacheFhevmSolidityConfigUnix() {
        const abs = this.cacheFhevmSolidityConfig;
        return (0, path_1.toUnixRelPath)(path.relative(this.root, abs));
    }
    get cacheFHEVMConfigSol() {
        return path.join(this.cacheFhevmSolidityConfig, constants_1.default.FHEVM_CONFIG_SOLIDITY_FILE);
    }
    get fhevmSolidity() {
        return path.dirname(this._resolveFromConsumer(path.join(constants_1.default.FHEVM_SOLIDITY_PACKAGE.name, "package.json")));
    }
    get solidityCoverage() {
        try {
            return path.dirname(this._resolveFromConsumer(path.join(constants_1.default.SOLIDITY_COVERAGE_PACKAGE_NAME, "package.json")));
        }
        catch {
            return undefined;
        }
    }
    get mockUtils() {
        try {
            return path.dirname(this._resolveFromConsumer(path.join(constants_1.default.FHEVM_MOCK_UTILS_PACKAGE_NAME, "package.json")));
        }
        catch {
            return undefined;
        }
    }
    get fhevmSolidityConfig() {
        return path.join(this.fhevmSolidity, path.dirname(constants_1.default.FHEVM_SOLIDITY_PACKAGE.configFile));
    }
    get fhevmSolidityConfigFile() {
        return path.join(this.fhevmSolidity, constants_1.default.FHEVM_SOLIDITY_PACKAGE.configFile);
    }
    get fhevmSolidityLib() {
        return path.join(this.fhevmSolidity, "lib");
    }
    get zamaFheOracleSolidity() {
        return path.dirname(this._resolveFromConsumer(path.join(constants_1.default.ZAMA_FHE_ORACLE_SOLIDITY_PACKAGE.name, "package.json")));
    }
    get zamaFheOracleSolidityAddress() {
        return path.join(this.zamaFheOracleSolidity, path.dirname(constants_1.default.ZAMA_FHE_ORACLE_SOLIDITY_PACKAGE.addressFile));
    }
    get cacheZamaFheOracleSolidity() {
        return path.join(this.cache, constants_1.default.ZAMA_FHE_ORACLE_SOLIDITY_PACKAGE.name);
    }
    get cacheZamaFheOracleSolidityAddress() {
        return path.join(this.cacheZamaFheOracleSolidity, path.dirname(constants_1.default.ZAMA_FHE_ORACLE_SOLIDITY_PACKAGE.addressFile));
    }
    get cacheZamaOracleAddressSol() {
        return path.join(this.cacheZamaFheOracleSolidity, constants_1.default.ZAMA_FHE_ORACLE_SOLIDITY_PACKAGE.addressFile);
    }
    get relCacheZamaFheOracleSolidityAddress() {
        const abs = this.cacheZamaFheOracleSolidityAddress;
        return path.relative(this.root, abs);
    }
    get relCacheZamaFheOracleSolidityAddressUnix() {
        const abs = this.cacheZamaFheOracleSolidityAddress;
        return (0, path_1.toUnixRelPath)(path.relative(this.root, abs));
    }
    resolveFhevmCoreContractsArtifactRootDir() {
        let modulePath = path.resolve(path.join("artifacts", constants_1.default.FHEVM_CORE_CONTRACTS_PACKAGE.name));
        if (!fs.existsSync(modulePath)) {
            modulePath = path.join(constants_1.default.FHEVM_CORE_CONTRACTS_PACKAGE.name, "artifacts");
        }
        return modulePath;
    }
    resolveZamaFheOracleSolidityArtifactRootDir() {
        let modulePath = path.resolve(path.join("artifacts", constants_1.default.ZAMA_FHE_ORACLE_SOLIDITY_PACKAGE.name));
        if (!fs.existsSync(modulePath)) {
            modulePath = path.join(constants_1.default.ZAMA_FHE_ORACLE_SOLIDITY_PACKAGE.name, "artifacts");
        }
        return modulePath;
    }
    resolveFhevmCoreContractsArtifactPath(contractName) {
        const root = this.resolveFhevmCoreContractsArtifactRootDir();
        return path.join(root, `contracts/${contractName}.sol/${contractName}.json`);
    }
    resolveZamaFheOracleSolidityArtifactPath(contractName) {
        const root = this.resolveZamaFheOracleSolidityArtifactRootDir();
        return path.join(root, `contracts/${contractName}.sol/${contractName}.json`);
    }
    async getFhevmCoreContractsArtifact(contractName) {
        const modulePath = this.resolveFhevmCoreContractsArtifactPath(contractName);
        const artifact = await Promise.resolve(`${modulePath}`).then(s => __importStar(require(s)));
        return { artifact, path: modulePath };
    }
    async getZamaFheOracleSolidityArtifact(contractName) {
        const modulePath = this.resolveZamaFheOracleSolidityArtifactPath(contractName);
        const artifact = await Promise.resolve(`${modulePath}`).then(s => __importStar(require(s)));
        return { artifact, path: modulePath };
    }
    getMockUtilsVersion() {
        try {
            const dir = this.mockUtils;
            if (!dir) {
                return undefined;
            }
            const pkgJson = JSON.parse(fs.readFileSync(path.join(dir, "package.json"), "utf8"));
            return pkgJson.version;
        }
        catch {
            return undefined;
        }
    }
    getBundledMockUtilsVersion() {
        return mock_utils_1.version;
    }
    _checkMockUtilsVersions() {
        const projectVersion = this.getMockUtilsVersion();
        if (!projectVersion) {
            return;
        }
        const bundledVersion = this.getBundledMockUtilsVersion();
        if (bundledVersion !== projectVersion) {
            throw new error_1.HardhatFhevmError(`Version mismatch detected for @fhevm/mock-utils.\n` +
                `> Installed in your project: @fhevm/mock-utils:${projectVersion}\n` +
                `> Expected (plugin): @fhevm/mock-utils:${bundledVersion}\n\n` +
                `Please ensure that your project is using the same version of @fhevm/mock-utils.\n` +
                `You can either:\n` +
                `- Align the versions by updating your dependencies\n` +
                `- Rely solely on the version provided by @fhevm/hardhat-plugin (no direct install)\n\n` +
                `This mismatch may lead to subtle runtime issues due to type incompatibilities or conflicting behavior.`);
        }
    }
    _resolveFromConsumer(modulePathId) {
        return resolveFromConsumer(modulePathId, this._root);
    }
}
exports.FhevmEnvironmentPaths = FhevmEnvironmentPaths;
function resolveFromConsumer(modulePathId, basedir) {
    try {
        const resolved = resolve.sync(modulePathId, {
            basedir,
        });
        return resolved;
    }
    catch {
        throw new error_1.HardhatFhevmError(`Unable to resolve ${modulePathId} from project at ${basedir}`);
    }
}
